File: solutions/216/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sun Oct 31 15:06:56 2021
> Description:   
 ************************************************************************/
#include <iostream> 
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
private:
    vector<vector<int>> result; 
    vector<int> path; 
public:
    void backtracking(int k, int n, int startId) {
        // prune invalid sum 
        if (n < 0) return; 
        // base case when we reach leaf node 
        if (k == 0) {
            if (n == 0) result.push_back(path); 
            return; 
        }
        // prune to make sure we can pick k numbers 
        // note here is 9-k+1 instead of 9-(k-path.size())+1
        for (int i = startId; i <= 9-k+1; ++i) {
            path.push_back(i); 
            backtracking(k-1, n-i, i+1); 
            path.pop_back(); 
        }
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        backtracking(k, n, 1); 
        return result; 
    }
};

TEST(Test216, SimpleTest) {
    const int k = 3; 
    const int n = 7; 
    Solution s; 
    vector<vector<int>> result = s.combinationSum3(k, n);
    for (const auto& m : result) {
        cout << "vector ";
        for (const auto& n : m)
            cout << n << " "; 
        cout << endl; 
    }
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/216/main.py
class Solution:
    def __init__(self): 
        self.path = []
        self.result = []
    def backtracking(self, k, n, startId): 
        if (n < 0): 
            return 
        if (k == 0):
            if (n == 0):
                # note here is self.path[:]
                self.result.append(self.path[:])
            return 
        # note here is 9 - k + 2, not 9 - (k - path.size()) + 2
        for i in range(startId, 9-k+2):
            self.path.append(i)
            self.backtracking(k-1, n-i, i+1)
            self.path.pop()
    def combinationSum3(self, k, n):
        self.backtracking(k, n, 1)
        return self.result 

if __name__ == "__main__":
    k, n = 3, 7
    s = Solution()
    print(f"result is {s.combinationSum3(k, n)}")
